data/tutorials/tt_02_file_manipulation.md

```diff
@@ -9,123 +9,618 @@ date: 2021-05-27T21:07:30-00:00
-  (* write something *)
-  close_out oc;
+In OCaml, these operations are handled through channels - abstract
+representations of input/output devices, and the OCaml standard library provides
+an array of functions to work with these channels.
+
+In this tutorial, we will cover how to perform basic file operations in OCaml.
+You will learn how to open and close channels, use these channels to read from
+and write to files and interact with the filesystem. We'll also share some
+important 'gotchas' to keep in mind while working with files in OCaml and how to
+handle common user errors and exceptions.
+
+To make the most of this tutorial, you should have some basic knowledge of
+OCaml, including an [understanding of types, functions and values](data-types).
+You should also have a basic understanding of how files work in the operating
+system.
+
+## Understanding Buffered Channels in OCaml
+
+Before we dive into file operations, it's essential to grasp the concept of
+"channels" and how they are used in OCaml.
+
+In computer programming, **channels** are an abstract representation of input or
+output devices. Think of them as the communication bridge between your program
+and external files or devices.
+
+A **buffered channel** is a type of channel that temporarily stores data in
+memory before it is written to or read from a file on disk. This can improve
+performance by reducing the number of disk accesses.
+
+In OCaml, buffered channels form the basis of file manipulation. There are two
+types of channels:
+
+1. **`out_channel`:** This type of channel is used when you want to write to a
+   file or an output device. It essentially represents a stream of data that
+   your program sends out.
+2. **`in_channel`:** Conversely, when you want to read from a file or an input
+   device, you would use an `in_channel`. It represents a stream of incoming
+   data that your program can process.
+
+A key thing to note is that when you open a file for reading or writing in
+OCaml, the function doesn't return a file descriptor like in some other
+languages. Instead, it returns a channel that you can read from or write to.
+
+In OCaml, standard process streams are also represented as channels. There are
+two output channels of type `out_channel`:
+- `stdout`: standard output, usually the console;
+- `stderr`: standard error output.
+
+There is an input channel, of type `in_channel` called `stdin`, which is the
+standard input, typically keyboard input.
+
+Now that you have a basic understanding of what channels are and their types, we
+can start exploring how to read from and write to files using these channels.
+
+## Writing to a File
+
+Here's a complete example of writing to a file in OCaml:
+
+```ocaml
+let write_to_file () =
+  (* Step 1: Open the File *)
+  let oc = Out_channel.open_text "file.txt" in
+  (* Step 2: Write to the Channel *)
+  Out_channel.output_string oc "Hello, World!\n";
+  (* Step 3: Flush the Channel *)
+  Out_channel.flush oc;
```

Comment:
Flush is not needed. It's needed in advanced use cases that generally don't involve files on the disk. I suggest removing it from this section. It is mentioned below in the "Common Pitfalls".